src/nginx.c

```diff
@@ -73,108 +85,261 @@ static size_t nginx_curl_callback(void *buf, size_t size, size_t nmemb,
   return len;
 }
 
-static int config_set(char **var, const char *value) {
-  if (*var != NULL) {
-    free(*var);
-    *var = NULL;
+/* Zeroing per nginx instance temporary stringlist for getting settings from
+ * collectd config for each nginx instance */
+static void init_nginx_instance_settings_array(void) {
+  for (int i = 0; i < config_keys_num; i++) {
+    for (int j = 0; j < DATA_MAX_NAME_LEN; j++) {
+      parsed_values_for_nginx_instance[i][j] = '\0';
+    }
   }
-
-  if ((*var = strdup(value)) == NULL)
-    return 1;
-  else
-    return 0;
 }
-
-static int config(const char *key, const char *value) {
-  if (strcasecmp(key, "url") == 0)
-    return config_set(&url, value);
-  else if (strcasecmp(key, "user") == 0)
-    return config_set(&user, value);
-  else if (strcasecmp(key, "password") == 0)
-    return config_set(&pass, value);
-  else if (strcasecmp(key, "verifypeer") == 0)
-    return config_set(&verify_peer, value);
-  else if (strcasecmp(key, "verifyhost") == 0)
-    return config_set(&verify_host, value);
-  else if (strcasecmp(key, "cacert") == 0)
-    return config_set(&cacert, value);
-  else if (strcasecmp(key, "timeout") == 0)
-    return config_set(&timeout, value);
-  else if (strcasecmp(key, "socket") == 0)
-    return config_set(&sock, value);
-  else
-    return -1;
-} /* int config */
-
-static int init(void) {
-  if (curl != NULL)
-    curl_easy_cleanup(curl);
-
-  if ((curl = curl_easy_init()) == NULL) {
-    ERROR("nginx plugin: curl_easy_init failed.");
-    return -1;
+/******************/
+
+/* Cleanup per nginx instance temporary stringlist for storing settings from
+ * collectd config for each nginx instance  */
+static void deinit_nginx_instance_settings_array(void) {
+  for (int i = 0; i < config_keys_num; i++) {
+    if (parsed_values_for_nginx_instance[i] != NULL) {
+      sfree(parsed_values_for_nginx_instance[i]);
+      parsed_values_for_nginx_instance[i] = NULL;
+    }
```

Comment:
While using an array makes certain things easier, I do not really like how you did it.  You have no names for the indexes, they're just magic values, when you use them later.  I.e. it's really annoying to check whether they e.g. match.

I'd rather have  `parsed_values_for_nginx_instance` a struct, with named members matching to config keys, so you can refer to them by name in `set_optional_curl_params()`.

Helper functions doing init + deinit for those struct members, could have auxiliary array though, like this:
`static const char *nginx_inst_str_ptrs[config_keys_num] = { &parsed_values_for_nginx_instance.username, ... };`

(Such helper could also include the names of the keys, so you can use it in parsing too.)